### PR TITLE
Added diou and ciou losses for bbox regression

### LINKED ISSUE #1085: GIOU, DIOU,CIOU
Hello! detectron2 use IOU loss, that is, smooth_l1_loss function.  will you plan to support of GIOU,DIOU,CIOU in detectron2?

### PR DESCRIPTION
**Pull request by SamFC10**
Resolves #1085

references and credits:
https://github.com/Zzh-tju/DIoU-pytorch-detectron/blob/master/lib/utils/net.py
https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/giou_loss.py

Thanks for your contribution!

If you're sending a large PR (e.g., >100 lines),
please open an issue first about the feature / bug, and indicate how you want to contribute.

We do not always accept features.
See https://detectron2.readthedocs.io/notes/contributing.html#pull-requests about how we handle PRs.

Before submitting a PR, please run `dev/linter.sh` to lint the code.

### CODE REVIEW

**Comment by SamFC10** [comment]
Hi @ppwwyyxx. Is there something more that needs to be added to this PR, like some tests? From my experiments, using diou/ciou loss instead of the default ones leads to better predictions and mAP, so it would be great to see these losses built-in to detectron2.

**Comment by ppwwyyxx** [comment]
Hi sorry for the delay. I didn't much time to look at big PRs.
The code looks correct but added some comments about style.

Please also move it to `layers/losses.py`.

Finally we'll need some unittests like those in `test_giou_loss.py` in fvcore. Testing for just a few special cases would be fine.

**Comment by SamFC10** [comment]
@ppwwyyxx  Moved losses.py to layers directory and added unit-tests for the 2 losses.

---

> The code looks correct but added some comments about style

I didn't get you. Where exactly are the comments about code-style?

**Review by ppwwyyxx — COMMENTED** [review]
oops the comments were not sent succesfully.

**Comment by SamFC10** [comment]
Addressed all code-style comments in latest commit. Will squash commits into one at the end if needed.

File: detectron2/modeling/losses.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                 'mean': The output will be averaged.
+                 'sum': The output will be summed.
+        eps (float): small number to prevent division by zero
+    """
+
+    x1, y1, x2, y2 = boxes1.unbind(dim=-1)
+    x1g, y1g, x2g, y2g = boxes2.unbind(dim=-1)
+
+    assert (x2 >= x1).all(), "bad box: x1 larger than x2"
+    assert (y2 >= y1).all(), "bad box: y1 larger than y2"
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply detectron2/modeling/losses.py]: CI failure : `AttributeError: module 'torch' has no attribute '_assert_async'`

[the remark this answers is not in the document]
Comment (ppwwyyxx) [reply detectron2/modeling/losses.py]: ok please still use `assert` but add a TODO that it should switch to `_assert_async` when we stop supporting pytorch1.8

[the remark this answers is not in the document]
Comment (SamFC10) [reply detectron2/modeling/losses.py]: Done

**Comment by SamFC10** [comment]
Hi @ppwwyyxx. Gentle ping. Is the PR ready to be merged?

**Pull request closed without merging**